PRO-3123 Make bill status code script triggerable by cronjob and wire in rails apis

shaun-prolegis opened the pull request:
## Why
In previous work, we set up a new EC2 instance to clone this repository (`Prolegis/congress`). We also introduced a script, `generate_bill_status_codes.sh`, which leverages UnitedStatesIO’s scripts to calculate bill status codes and upload them to S3.

To finalize this setup, the following tasks are needed:
- Update the script so it can be executed by a cron job.
- Use the Rails API to fetch the current Congress dynamically, rather than hardcoding it as 118.
- Enable the bill status code importer for staging, demo, and production environments.

## How
To enable the script to run via cron, I added the following line to ensure the correct directory is used:  
`cd ~/congress || exit`

This is necessary because cron jobs use a single configuration file for the entire EC2 instance. There is no separate configuration file specifically for the congress repository, so we need to guarantee the script runs from the appropriate directory, whether manually or through the cron job.

Additionally, I updated the script to fetch the current Congress programmatically from the Rails API, eliminating the need for hardcoding. Finally, I added cURL commands to trigger the bill status code importer across staging, demo, and production environments.

shaun-prolegis committed:
Make bill status code script triggerable by cronjob and wire in rails apis

shaun-prolegis commented on generate_bill_status_codes.sh:
```diff
@@ -1,20 +1,19 @@
-CURRENT_CONGRESS=118
-echo "The current congress is: $CURRENT_CONGRESS"
+cd ~/congress || exit # Change into the congress directory, if not already in it.
+git pull # Fetch the latest version of the congress repository from GitHub
```
I also added in this so that we are always getting the latest version.

shaun-prolegis commented on generate_bill_status_codes.sh:
```diff
@@ -1,20 +1,19 @@
-# For now, we will store it as a hard-coded variable
-CURRENT_CONGRESS=118
-echo "The current congress is: $CURRENT_CONGRESS"
+cd ~/congress || exit # Change into the congress directory, if not already in it.
```
The `|| exit` is a safety check in the case that this directory doesn't exist ( my IDE recommended it to me). If the directory doesn't exist, the script terminates.

harish-prolegis replied:
why exit? why can't you run `git clone` ? that would create the directory right? and then you can go on to do the rest.

shaun-prolegis commented on .gitignore:
```diff
@@ -57,3 +57,7 @@ docs/_build/
 # Local things
 /dev/
+
+# Stuff from Prolegis import scripts
```
Since these are files that are being generated either locally on a developer's machine, or on the EC2 instance, I don't think that they should be included in version control, and we should add them to this `.gitignore`.

harish-prolegis reviewed (commented):
Few questions

1. We are unfortunately only counting for positive scenario. We should consider for negative scenarios as well. For instance, what if the EC2 instance goes down? Or what if the requests to prolegis doesn't go through? Should we not safe guard against these scenarios?
2. We should get notified when the script finishes running, so that we all know that it has run the script. In case we don't receive any notifications, we will know to look into the script.

shaun-prolegis commented:
Hi Harish! Thanks for the great questions. I dug into the concerns you raised, and I’ve implemented some changes that I think address the points you mentioned, specifically adding logging, error handling, and considering failure scenarios such as the EC2 instance going down or requests to Prolegis failing.

I’ve created a [new PR](https://github.com/Prolegis/congress/pull/4/files) with the proposed changes. Since the changes were quite significant, I decided to branch off of this feature branch for testing and experimenting purposes. Once everything is stable, we can merge these changes back into this feature branch.

### New Endpoints:

I introduced two new Rails endpoints in the approach to improve logging and validation:

- **`POST /api/congress_repo/log_message`**: This allows us to log various events and errors during the import process.
- **`GET /api/congress_repo/validate_key`**: This endpoint helps us validate API keys for each environment (production, staging, and demo) to ensure they're working before proceeding with the script.

We would need to add these in before add this script in. 
Here’s an example of how these controllers would look in our Rails app:

```ruby
class Api::CongressRepo::LogMessageController < Api::CongressRepo::BaseController
  def create
    message = params[:message]
    level = params[:level] || "info" # Default log level is "info"
    Log.log(message, level: level, tag: self.class.name)
    head :ok
  end
end
```

```ruby
class Api::CongressRepo::ValidateKeyController < Api::CongressRepo::BaseController
  def index
    head :ok
  end
end
```

The `validate_key` endpoint ensures that the API keys we are using on the EC2 instance are valid. If not, the script will halt and log an error. 

### Script Changes

In the script, I added logic to validate the API keys before proceeding with the import. If the validation fails, we log an error and stop the process to prevent any further issues. Additionally, the script now logs both the start and completion of the import, as well as any errors that may occur (e.g., cloning the repo, AWS S3 copy failure).

Here’s the modified script:

```bash
#!/bin/bash

log_file="bill_status_codes_import_log.txt" # Log file
echo "Bill Status Codes Import started at $(date '+%Y-%m-%d %H:%M:%S')" >> $log_file

# Validate Production Key
response=$(curl -s -o /dev/null -w "%{http_code}" -H "Authorization: $API_KEY_PRODUCTION" https://www.prolegis.com/api/congress_repo/validate_key)
if [ "$response" -ne 200 ]; then
  echo "Production key is invalid. HTTP status: $response" >> $log_file
  exit 1
fi

# Validate Staging Key
response=$(curl -s -o /dev/null -w "%{http_code}" -H "Authorization: $API_KEY_STAGING" https://stg.prolegis.com/api/congress_repo/validate_key)
if [ "$response" -ne 200 ]; then
  echo "Staging key is invalid. HTTP status: $response" >> $log_file
  exit 1
fi

# Validate Demo Key
response=$(curl -s -o /dev/null -w "%{http_code}" -H "Authorization: $API_KEY_DEMO" https://demo.prolegis.com/api/congress_repo/validate_key)
if [ "$response" -ne 200 ]; then
  echo "Demo key is invalid. HTTP status: $response" >> $log_file
  exit 1
fi

# Send a log message that the import has started
response=$(curl -s -o /dev/null -w "%{http_code}" -X POST https://www.prolegis.com/api/congress_repo/log_message \
-H "Authorization: $API_KEY_PRODUCTION" \
-H "Content-Type: application/json" \
-d "{\"message\": \"Bill Status Codes Import started at $(date '+%Y-%m-%d %H:%M:%S')\", \"level\": \"info\"}")

if [ "$response" -ne 200 ]; then
  echo "Failed to log start of import. HTTP status: $response" >> $log_file
  exit 1
fi

# Clone from GitHub if not already cloned
git clone "https://$GITHUB_PERSONAL_ACCESS_TOKEN@github.com/Prolegis/congress.git"
if [ $? -ne 0 ]; then
  echo "Git clone failed, logging the error." >> $log_file
  curl -X POST https://www.prolegis.com/api/congress_repo/log_message \
  -H "Authorization: $API_KEY_PRODUCTION" \
  -H "Content-Type: application/json" \
  -d "{\"message\": \"Git clone failed for the congress repository\", \"level\": \"error\"}"
  exit 1
fi

cd congress
git pull

output_file="bill_status_codes.json" # Output JSON file
echo "Bill Status Codes Import started at $(date '+%Y-%m-%d %H:%M:%S')" >> $log_file

# Fetch the current congress from Rails API
CURRENT_CONGRESS=$(curl -s -H "Authorization: $API_KEY_PRODUCTION" https://www.prolegis.com/api/congress_repo/current_congress | jq -r '.current_congress')

# Proceed with the rest of the import process...
```

### Summary

These changes improve the reliability of the script by:
- Adding API key validation to ensure we’re operating with valid credentials.
- Logging the start and completion of the import, as well as any errors, to both a log file and via an API endpoint.
- Enhancing error handling for critical steps like the Git clone operation.

Let me know what you think about this approach, and we can iterate further as needed. Thanks! :D

harish-prolegis reviewed (approved):
lgtm
